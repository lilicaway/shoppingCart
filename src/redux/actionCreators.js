import * as actionType from './actions';
import axios from 'axios';

export const addPromotion = promotion => {
  // promotion has these fields:
  // { id, name, type, discountPercent?, forQuantity?, pay? }
  // Where the last 3 depend on the type.
  return { type: actionType.ADD_PROMOTION, payload: promotion };
};

export const addProduct = product => {
  // product has these fields:
  // { id, name, price, promotionId?}
  return { type: actionType.ADD_PRODUCT, payload: product };
};

export const addCartItem = productId => {
  return { type: actionType.ADD_CART_ITEM, payload: { productId } };
};

export const incrementCartItemQuantity = productId => {
  return { type: actionType.INCREMENT_CART_ITEM_QUANTITY, payload: { productId } };
};

export const decrementCartItemQuantity = productId => {
  return { type: actionType.DECREMENT_CART_ITEM_QUANTITY, payload: { productId } };
};

export const setDataAsLoading = () => {
  return { type: actionType.DATA_LOADER_LOADING, payload: undefined };
};

export const setDataAsError = error => {
  return { type: actionType.DATA_LOADER_ERROR, payload: error };
};

export const setDataAsCompleted = () => {
  return { type: actionType.DATA_LOADER_COMPLETED, payload: undefined };
};

export const loadData = () => {
  return dispatch => {
    dispatch(setDataAsLoading());
    axios
      .get('/promotions.json')
      .then(response => {
        response.data.map(promotion => dispatch(addPromotion(promotion)));
      })
      .then(() => {
        // Products depend on promotions. That is why we load them serialy.
        return axios.get('/products.json').then(response => {
          response.data.map(product => dispatch(addProduct(product)));
        });
      })
      .then(() => {
        dispatch(setDataAsCompleted());
      })
      .catch(error => {
        dispatch(setDataAsError(error.toString()));
      });
  };
};
