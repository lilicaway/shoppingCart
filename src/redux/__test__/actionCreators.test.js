import moxios from 'moxios';
import * as actionCreator from '../actionCreators';

test('addPromotion', () => {
  expect(
    actionCreator.addPromotion({
      id: '1d7bae3b-593c-4335-8b59-1a9b6031b47b',
      name: '10% discount',
      type: 'PERCENTAGE',
      discountRate: 0.1
    })
  ).toMatchSnapshot();
});

test('addProduct', () => {
  expect(
    actionCreator.addProduct({
      id: '42d1595f-63d6-4c68-967a-e1ec4d7ca37c',
      name: 'Apple',
      price: 0.25
    })
  ).toMatchSnapshot();
});

test('addCartItem', () => {
  expect(actionCreator.addCartItem('42d1595f-63d6-4c68-967a-e1ec4d7ca37c')).toMatchSnapshot();
});

test('incrementCartItemQuantity', () => {
  expect(actionCreator.incrementCartItemQuantity('42d1595f-63d6-4c68-967a-e1ec4d7ca37c')).toMatchSnapshot();
});

test('decrementCartItemQuantity', () => {
  expect(actionCreator.decrementCartItemQuantity('42d1595f-63d6-4c68-967a-e1ec4d7ca37c')).toMatchSnapshot();
});

test('setDataAsLoading', () => {
  expect(actionCreator.setDataAsLoading()).toMatchSnapshot();
});

test('setDataAsError', () => {
  expect(actionCreator.setDataAsError('some error message')).toMatchSnapshot();
});

test('setDataAsCompleted', () => {
  expect(actionCreator.setDataAsCompleted()).toMatchSnapshot();
});

test('getAPIDetails', done => {
  const quantityPromotion = {
    id: '85d80dfa-46b9-4e87-8a76-2ad2cfcca040',
    name: '3 x 2',
    type: 'QUANTITY',
    forQuantity: 3,
    pay: 2
  };
  const expectedPromotions = [quantityPromotion];

  const dispatchMock = jest.fn();
  moxios.withMock(() => {
    actionCreator.loadData()(dispatchMock);

    moxios.wait(() => {
      try {
        expect(dispatchMock).toBeCalledWith(actionCreator.setDataAsLoading());
        const promotionsRequest = moxios.requests.mostRecent();
        promotionsRequest
          .respondWith({ status: 200, response: expectedPromotions })
          .then(() => {
            expect(promotionsRequest.url).toEqual('/promotions.json');
            expect(dispatchMock).toBeCalledWith(actionCreator.addPromotion(quantityPromotion));
            done();
          })
          .catch(done.fail); // https://jasmine.github.io/2.3/introduction.html#section-54
      } catch (e) {
        done.fail(e);
      }
    });
  });
});
