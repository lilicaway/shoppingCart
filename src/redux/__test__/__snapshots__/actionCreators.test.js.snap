// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addCartItem 1`] = `
Object {
  "payload": Object {
    "productId": "42d1595f-63d6-4c68-967a-e1ec4d7ca37c",
  },
  "type": "ADD_CART_ITEM",
}
`;

exports[`addProduct 1`] = `
Object {
  "payload": Object {
    "id": "42d1595f-63d6-4c68-967a-e1ec4d7ca37c",
    "name": "Apple",
    "price": 0.25,
  },
  "type": "ADD_PRODUCT",
}
`;

exports[`addPromotion 1`] = `
Object {
  "payload": Object {
    "discountRate": 0.1,
    "id": "1d7bae3b-593c-4335-8b59-1a9b6031b47b",
    "name": "10% discount",
    "type": "PERCENTAGE",
  },
  "type": "ADD_PROMOTION",
}
`;

exports[`decrementCartItemQuantity 1`] = `
Object {
  "payload": Object {
    "productId": "42d1595f-63d6-4c68-967a-e1ec4d7ca37c",
  },
  "type": "DECREMENT_CART_ITEM_QUANTITY",
}
`;

exports[`incrementCartItemQuantity 1`] = `
Object {
  "payload": Object {
    "productId": "42d1595f-63d6-4c68-967a-e1ec4d7ca37c",
  },
  "type": "INCREMENT_CART_ITEM_QUANTITY",
}
`;

exports[`setDataAsCompleted 1`] = `
Object {
  "payload": undefined,
  "type": "DATA_LOADER_COMPLETE",
}
`;

exports[`setDataAsError 1`] = `
Object {
  "payload": "some error message",
  "type": "DATA_LOADER_ERROR",
}
`;

exports[`setDataAsLoading 1`] = `
Object {
  "payload": undefined,
  "type": "DATA_LOADER_LOADING",
}
`;
